File: sdk/src/nomabot/render/background_cache.py
"""BackgroundCache patch restore logic (Python mirror for tests)."""

from __future__ import annotations


class BackgroundCache:
    """Stores an RGB565 background patch for restore-before-blit."""

    def __init__(self) -> None:
        self._pixels: list[int] = []
        self._x = 0
        self._y = 0
        self._w = 0
        self._h = 0
        self._valid = False

    @property
    def valid(self) -> bool:
        return self._valid

    def reset(self) -> None:
        self._valid = False
        self._w = 0
        self._h = 0

    def capture_from_background(
        self,
        bg_pixels: list[int],
        bg_w: int,
        bg_h: int,
        patch_x: int,
        patch_y: int,
        patch_w: int,
        patch_h: int,
    ) -> bool:
        if not bg_pixels or bg_w <= 0 or bg_h <= 0 or patch_w <= 0 or patch_h <= 0:
            self.reset()
            return False

        src_x = patch_x
        src_y = patch_y
        if src_x < 0:
            patch_w += src_x
            src_x = 0
        if src_y < 0:
            patch_h += src_y
            src_y = 0
        if src_x + patch_w > bg_w:
            patch_w = bg_w - src_x
        if src_y + patch_h > bg_h:
            patch_h = bg_h - src_y
        if patch_w <= 0 or patch_h <= 0:
            self.reset()
            return False

        pixels: list[int] = []
        for row in range(patch_h):
            row_start = (src_y + row) * bg_w + src_x
            pixels.extend(bg_pixels[row_start : row_start + patch_w])

        self._pixels = pixels
        self._x = src_x
        self._y = src_y
        self._w = patch_w
        self._h = patch_h
        self._valid = True
        return True

    def restore(self, canvas: list[list[int]]) -> None:
        if not self._valid or self._w <= 0 or self._h <= 0:
            return
        idx = 0
        for row in range(self._h):
            y = self._y + row
            if y < 0 or y >= len(canvas):
                idx += self._w
                continue
            row_data = canvas[y]
            for col in range(self._w):
                x = self._x + col
                if 0 <= x < len(row_data):
                    row_data[x] = self._pixels[idx]
                idx += 1

File: firmware/src/render/background_cache.cpp
#include "background_cache.h"

#include <Arduino.h>
#include <cstring>

BackgroundCache::~BackgroundCache() {
  if (_pixels) {
    free(_pixels);
    _pixels = nullptr;
  }
}

void BackgroundCache::reset() {
  _valid = false;
  _w = 0;
  _h = 0;
}

bool BackgroundCache::ensureCapacity(size_t pixelCount) {
  if (pixelCount == 0) {
    return false;
  }
  if (_capacity >= pixelCount && _pixels) {
    return true;
  }
  uint16_t *next = static_cast<uint16_t *>(realloc(_pixels, pixelCount * sizeof(uint16_t)));
  if (!next) {
    return false;
  }
  _pixels = next;
  _capacity = pixelCount;
  return true;
}

bool BackgroundCache::captureFromBackground(const uint16_t *bgPixels, int bgW, int bgH,
                                            int patchX, int patchY, int patchW, int patchH) {
  if (!bgPixels || bgW <= 0 || bgH <= 0 || patchW <= 0 || patchH <= 0) {
    _valid = false;
    return false;
  }

  int srcX = patchX;
  int srcY = patchY;
  if (srcX < 0) {
    patchW += srcX;
    srcX = 0;
  }
  if (srcY < 0) {
    patchH += srcY;
    srcY = 0;
  }
  if (srcX + patchW > bgW) {
    patchW = bgW - srcX;
  }
  if (srcY + patchH > bgH) {
    patchH = bgH - srcY;
  }
  if (patchW <= 0 || patchH <= 0) {
    _valid = false;
    return false;
  }

  if (!ensureCapacity(static_cast<size_t>(patchW) * static_cast<size_t>(patchH))) {
    _valid = false;
    return false;
  }

  _x = srcX;
  _y = srcY;
  _w = patchW;
  _h = patchH;

  for (int row = 0; row < patchH; ++row) {
    const uint16_t *srcRow = bgPixels + (srcY + row) * bgW + srcX;
    uint16_t *dstRow = _pixels + row * patchW;
    memcpy(dstRow, srcRow, static_cast<size_t>(patchW) * sizeof(uint16_t));
  }

  _valid = true;
  return true;
}

void BackgroundCache::restore(IRenderer &renderer) const {
  if (!_valid || !_pixels || _w <= 0 || _h <= 0) {
    return;
  }
  renderer.blitRGB565(_pixels, _x, _y, _w, _h);
}
